{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "1b75ff6f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os, shutil\n",
    "import inspect\n",
    "import datetime as dt\n",
    "from sys import platform\n",
    "from glob import glob\n",
    "from tqdm import tqdm\n",
    "import openpyxl\n",
    "from openpyxl.utils.dataframe import dataframe_to_rows\n",
    "\n",
    "import seronetDataclass as seroClass\n",
    "import seronetFunctions as seroFxn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "cf7a1fbb",
   "metadata": {},
   "outputs": [],
   "source": [
    "cloud = '/Users/liualg/Library/CloudStorage/OneDrive-NationalInstitutesofHealth/Curation channel/ImmPort Uploads/'\n",
    "PMID_LIST = pd.read_excel(os.path.join(cloud,'Immport-DR46 (November).xlsx'))['PMID (link to Box)'][1:]\n",
    "PMID_LIST = [\n",
    "    35180044, 35390296, 34368647, 35284808, 35483404, 34546094, 34250518, 34320281, 35085183, 35061630, 35704428, 33688034, 35756977, 34150933, 35090596, 35932763, 35839768, 35081612, 34383889, 36058184, 35576468, 34308390, 35881010, 35881005\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b9b3925e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# finding correct Box Base\n",
    "if platform == \"darwin\":\n",
    "    box_base = \"~/Library/CloudStorage/Box-Box/SeroNet Public Data\"\n",
    "else: \n",
    "    print(\"User has windows\")\n",
    "    box_base = os.path.join(\"Users\",os.getlogin(), \"Box\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a5da878",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "aa1e06a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_all_files(pmid):\n",
    "    if isinstance(pmid, int):\n",
    "        pmid = str(pmid)\n",
    "    else:\n",
    "        pass\n",
    "    #File Paths\n",
    "    BASE_DIR = seroFxn.get_box_dir(box_base, pmid)\n",
    "    OUT_DIR = glob(os.path.join(BASE_DIR,'ImmPort*'))\n",
    "    OUT_DIR.sort()\n",
    "    \n",
    "    for file in glob(os.path.join(OUT_DIR[-1],\"*.txt\")):\n",
    "        if 'protocol.txt' in file:\n",
    "            protocols.append(pd.read_csv(file,sep='\\t',header=2).fillna('-'))\n",
    "\n",
    "#         elif 'experiments' in file:\n",
    "#             experiments.append(pd.read_csv(file,sep='\\t',header=2).fillna('-'))\n",
    "\n",
    "        elif 'reagent' in file:\n",
    "            reagents.append(pd.read_csv(file,sep='\\t',header=2).fillna('-'))\n",
    "\n",
    "        elif 'panel' in file:\n",
    "            assessments.append(pd.read_csv(file,sep='\\t',header=2).fillna('-'))\n",
    "\n",
    "        elif 'subject_human' in file:\n",
    "            subject_humans.append(pd.read_csv(file,sep='\\t',header=2).fillna('-'))\n",
    "\n",
    "        elif 'subject_organism' in file:\n",
    "            subject_organisms.append(pd.read_csv(file,sep='\\t',header=2).fillna('-'))\n",
    "\n",
    "        elif 'experiment_samples' in file:\n",
    "            experiment_samples.append(pd.read_csv(file,sep='\\t',header=2).fillna('-'))\n",
    "\n",
    "        elif 'treatment' in file:\n",
    "            treatments.append(pd.read_csv(file,sep='\\t',header=2).fillna('-'))\n",
    "\n",
    "        elif 'basic' in file:\n",
    "            basic_stdy_templates.append(file)\n",
    "        else:\n",
    "#             print(OUT_DIR[-1])\n",
    "            pass\n",
    "\n",
    "#     return {'basic_stdy_templates':basic_stdy_templates, 'protocols': protocols, 'experiments': experiments,\n",
    "#             'reagents': reagents,'assessments': assessments,'subject_humans': subject_humans,\n",
    "#             'subject_organisms': subject_organisms, 'experiment_samples': experiment_samples,\n",
    "#             'treatments': treatments\n",
    "#            }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "a4c2a443",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_sections(ws, class_names):\n",
    "    temp = []\n",
    "    for i in range(1,ws.max_row):\n",
    "        if ws[\"A\"][i].value in class_names:\n",
    "            temp.append(i+1)\n",
    "    return temp\n",
    "\n",
    "def replace_header(dataframe):\n",
    "    dataframe.columns = dataframe.iloc[0]\n",
    "    return dataframe[1:]\n",
    "\n",
    "# def clean_df(dataframe, add_PMID = False):\n",
    "    \n",
    "#     dataframe = seroFxn.edit_df(dataframe)\n",
    "#     dataframe = dataframe.dropna(how='all',axis=0).dropna(how='all',axis=1)\n",
    "#     if add_PMID:\n",
    "#         dataframe['PMID'] = [PMID]*len(dataframe)\n",
    "#     else:\n",
    "#         pass\n",
    "    \n",
    "#     reut\n",
    "# #             df['PMID'] = [PMID]*len(df) #LIU\n",
    "# #             CAT.append(df)\n",
    "\n",
    "\n",
    "\n",
    "#     dataframe = seroFxn.edit_df(dataframe)\n",
    "#     dataframe = dataframe.dropna(how='all',axis=0).dropna(how='all',axis=1)\n",
    "#     if add_PMID:\n",
    "#         dataframe['PMID'] = [PMID]*len(dataframe)\n",
    "    \n",
    "#     return replace_header(dataframe.reset_index())\n",
    "\n",
    "def get_basic_info(ws, PMID):\n",
    "    # Class names\n",
    "    class_names = ['study', 'study_personnel', 'study_categorization', 'study_2_condition_or_disease',\n",
    "                       'arm_or_cohort', 'study_personnel', 'planned_visit',\n",
    "                       'inclusion_exclusion', 'study_2_protocol', 'study_file',\n",
    "                       'study_link','study_pubmed'\n",
    "                  ]\n",
    "\n",
    "\n",
    "\n",
    "    VARS_TO_CLEAN = ['', 'N/A', 'n/a', 'na', np.nan, None]\n",
    "    clean_other = VARS_TO_CLEAN + ['Other']\n",
    "\n",
    "    ws.delete_cols(1)  \n",
    "    sp = get_sections(ws, class_names)\n",
    "    sp.append(200)\n",
    "    \n",
    "    for section_number in range(len(sp)-1):\n",
    "        temp_wb = openpyxl.Workbook()\n",
    "        temp_ws = temp_wb.active\n",
    "\n",
    "        #making a temp workbook to store each section. This will be turned into df\n",
    "        for i in ws.iter_rows(values_only = True,\n",
    "                                    min_row = sp[section_number]+1,\n",
    "                                    max_row = sp[section_number+1]-1):\n",
    "            temp_ws.append(i)\n",
    "            \n",
    "#         max_row = temp_ws.max_row\n",
    "#         max_col = temp_ws.max_column\n",
    "        seroFxn.remove_excess(temp_ws)\n",
    "        \n",
    "        df = pd.DataFrame(temp_ws.values)\n",
    "        sub_section = ws.cell(row=sp[section_number], column = 1).value.strip()\n",
    "#         print(sub_section)\n",
    "        if sub_section == 'study':\n",
    "            df = seroFxn.edit_df(df)\n",
    "            df = df.dropna(how='all')\n",
    "            STUDY.append(df)\n",
    "            \n",
    "        elif sub_section == 'study_categorization':\n",
    "            df = seroFxn.edit_df(df)\n",
    "            df = df.dropna(how='all',axis=0).dropna(how='all',axis=1)\n",
    "            df['PMID'] = [PMID]*len(df) #LIU\n",
    "            CAT.append(df)\n",
    "            \n",
    "        elif sub_section == 'study_2_condition_or_disease':\n",
    "            df = seroFxn.edit_df(df)\n",
    "            df = df.dropna(how='all',axis=0).dropna(how='all',axis=1)\n",
    "            df['PMID'] = [PMID]*len(df)\n",
    "            COD.append(df)\n",
    "            \n",
    "        elif sub_section == 'inclusion_exclusion':\n",
    "            df = seroFxn.edit_df(df).T\n",
    "            df = df.dropna(how='all',axis=0).dropna(how='all',axis=1)\n",
    "            IE.append(replace_header(df.reset_index()))\n",
    "            \n",
    "        elif sub_section == 'arm_or_cohort':\n",
    "            df = seroFxn.edit_df(df).T\n",
    "            df = df.dropna(how='all',axis=0).dropna(how='all',axis=1)\n",
    "            AOC.append(replace_header(df.reset_index()))\n",
    "            \n",
    "        elif sub_section == 'study_personnel':\n",
    "            df = seroFxn.edit_df(df).T\n",
    "            df = df.dropna(how='all').dropna(how='all',axis=1)\n",
    "            PERSONNEL.append(replace_header(df.reset_index()))\n",
    "            \n",
    "        elif sub_section == 'planned_visit':\n",
    "            df = seroFxn.edit_df(df).T\n",
    "            df = df.dropna(how='all').dropna(how='all',axis=1)\n",
    "            PV.append(replace_header(df.reset_index()))\n",
    "            \n",
    "        elif sub_section == 'study_file':\n",
    "            df = seroFxn.edit_df(df).T\n",
    "            df = df.dropna(how='all').dropna(how='all',axis=1)\n",
    "            FILE.append(replace_header(df.reset_index()))\n",
    "            \n",
    "        elif sub_section == 'study_link':\n",
    "            df = seroFxn.edit_df(df).T\n",
    "            if df.empty:\n",
    "                pass\n",
    "            else:\n",
    "                df = df.dropna(how='all').dropna(how='all',axis=1)\n",
    "                df = replace_header(df)\n",
    "                df['PMID'] = [PMID]*len(df)\n",
    "                LINK.append(df.reset_index())\n",
    "            \n",
    "        else:\n",
    "            pass\n",
    "        \n",
    "#     return {\n",
    "#      'STUDY':STUDY,'CAT':CAT,'COD':COD,'IE':IE,'AOC':AOC,'PERSONNEL' :PERSONNEL,'PV' :PV,'FILE' :FILE, 'LINK' :LINK\n",
    "#     }\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "0530eee0",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|███████████████████████████████████████████| 24/24 [00:02<00:00,  9.35it/s]\n"
     ]
    }
   ],
   "source": [
    "#templates excluding study\n",
    "basic_stdy_templates = []\n",
    "protocols = []\n",
    "experiments = []\n",
    "reagents = []\n",
    "assessments = []\n",
    "subject_humans = []\n",
    "subject_organisms = []\n",
    "experiment_samples = []\n",
    "treatments = []\n",
    "\n",
    "for PMID in tqdm(PMID_LIST):\n",
    "    PMID = str(PMID)\n",
    "    get_all_files(PMID)\n",
    "\n",
    "# basic_stdy_template_report = []\n",
    "protocols_report = pd.concat(protocols)\n",
    "reagent_report = pd.concat(reagents)\n",
    "assessment_report = pd.concat(assessments)\n",
    "subject_human_report = pd.concat(subject_humans)\n",
    "subject_organism_report = pd.concat(subject_organisms)\n",
    "experiment_sample_report = pd.concat(experiment_samples)\n",
    "treatment_report = pd.concat(treatments)    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "015a91aa",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|███████████████████████████████████████████| 23/23 [00:01<00:00, 12.63it/s]\n"
     ]
    }
   ],
   "source": [
    "#study split\n",
    "STUDY  = []\n",
    "CAT = []\n",
    "COD = []\n",
    "IE = []\n",
    "AOC = []\n",
    "PERSONNEL = []\n",
    "PV = []\n",
    "FILE = []\n",
    "LINK = []\n",
    "\n",
    "for basic_path in tqdm(basic_stdy_templates):\n",
    "    basic_wb = openpyxl.Workbook()\n",
    "    basic_ws = basic_wb.active\n",
    "\n",
    "    for r in dataframe_to_rows(pd.read_csv(basic_path, sep='\\t',header=None)):\n",
    "        basic_ws.append(r)\n",
    "\n",
    "    get_basic_info(basic_ws, re.findall(r'\\d{8}',basic_path)[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "4e010048",
   "metadata": {},
   "outputs": [],
   "source": [
    "# basic_stdy_template_report = []\n",
    "\n",
    "protocols_report = pd.concat(protocols).reset_index(drop=True)\n",
    "reagent_report = pd.concat(reagents).reset_index(drop=True)\n",
    "assessment_report = pd.concat(assessments).reset_index(drop=True)\n",
    "subject_human_report = pd.concat(subject_humans).reset_index(drop=True)\n",
    "subject_organism_report = pd.concat(subject_organisms).reset_index(drop=True)\n",
    "experiment_sample_report = pd.concat(experiment_samples).reset_index(drop=True)\n",
    "treatment_report = pd.concat(treatments).reset_index(drop=True)\n",
    "\n",
    "STUDY_report = pd.concat(STUDY).dropna(axis=1,how='all').reset_index(drop=True)\n",
    "COD_report = pd.concat(COD).reset_index(drop=True)\n",
    "CAT_report = pd.concat(CAT).reset_index(drop=True)\n",
    "IE_report = pd.concat(IE).reset_index(drop=True)\n",
    "AOC_report = pd.concat(AOC).reset_index(drop=True)\n",
    "PERSONNEL_report = pd.concat(PERSONNEL).reset_index(drop=True)\n",
    "PV_report = pd.concat(PV).reset_index(drop=True)\n",
    "FILE_report = pd.concat(FILE).reset_index(drop=True)\n",
    "LINK_report = pd.concat(LINK).reset_index(drop=True)\n",
    "LINK_report.columns = ['Name','Value','PMID']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "dc4bf99e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Value</th>\n",
       "      <th>PMID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Gene Expression Omnibus (GEO)</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi...</td>\n",
       "      <td>35483404</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Gene Expression Omnibus (GEO)</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi...</td>\n",
       "      <td>35483404</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>BioProject</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/bioproject/?term=...</td>\n",
       "      <td>34546094</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRA</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/sra/?term=SRR1431...</td>\n",
       "      <td>34546094</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRA</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/sra/?term=SRR1431...</td>\n",
       "      <td>34546094</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>GenBank</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/nuccore/MN985325</td>\n",
       "      <td>33688034</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Github</td>\n",
       "      <td>https://github.com/boyercb/covid-gathering-size</td>\n",
       "      <td>36058184</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Metabolomics Workbench: Raw and analyzed data</td>\n",
       "      <td>https://doi.org/10.21228/M80981</td>\n",
       "      <td>34308390</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Machine Learning Analysis Code</td>\n",
       "      <td>https://github.com/e-stan/covid_19_analysis</td>\n",
       "      <td>34308390</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>NCBI SRA (PRJNA806767)</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/bioproject/PRJNA8...</td>\n",
       "      <td>35881005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>NCBI SRA (PRJNA802993)</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/bioproject/PRJNA8...</td>\n",
       "      <td>35881005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>Zenodo</td>\n",
       "      <td>https://zenodo.org/record/6585475#.Yvvjd3bMIuU</td>\n",
       "      <td>35881005</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>Zenodo</td>\n",
       "      <td>https://zenodo.org/record/6887187#.YvvkFHbMIuU</td>\n",
       "      <td>35881005</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             Name  \\\n",
       "0                   Gene Expression Omnibus (GEO)   \n",
       "1                   Gene Expression Omnibus (GEO)   \n",
       "2                                      BioProject   \n",
       "3                                             SRA   \n",
       "4                                             SRA   \n",
       "5                                         GenBank   \n",
       "6                                          Github   \n",
       "7   Metabolomics Workbench: Raw and analyzed data   \n",
       "8                  Machine Learning Analysis Code   \n",
       "9                          NCBI SRA (PRJNA806767)   \n",
       "10                         NCBI SRA (PRJNA802993)   \n",
       "11                                        Zenodo    \n",
       "12                                        Zenodo    \n",
       "\n",
       "                                                Value      PMID  \n",
       "0   https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi...  35483404  \n",
       "1   https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi...  35483404  \n",
       "2   https://www.ncbi.nlm.nih.gov/bioproject/?term=...  34546094  \n",
       "3   https://www.ncbi.nlm.nih.gov/sra/?term=SRR1431...  34546094  \n",
       "4   https://www.ncbi.nlm.nih.gov/sra/?term=SRR1431...  34546094  \n",
       "5       https://www.ncbi.nlm.nih.gov/nuccore/MN985325  33688034  \n",
       "6     https://github.com/boyercb/covid-gathering-size  36058184  \n",
       "7                     https://doi.org/10.21228/M80981  34308390  \n",
       "8         https://github.com/e-stan/covid_19_analysis  34308390  \n",
       "9   https://www.ncbi.nlm.nih.gov/bioproject/PRJNA8...  35881005  \n",
       "10  https://www.ncbi.nlm.nih.gov/bioproject/PRJNA8...  35881005  \n",
       "11     https://zenodo.org/record/6585475#.Yvvjd3bMIuU  35881005  \n",
       "12     https://zenodo.org/record/6887187#.YvvkFHbMIuU  35881005  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "LINK_report"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "52e65db5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# os.mkdir(\"./DR46/database_report/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "a079cbd6",
   "metadata": {},
   "outputs": [],
   "source": [
    "protocols_report.to_csv('./DR46/database_report/protocols_report.csv')\n",
    "reagent_report.to_csv('./DR46/database_report/reagent_report.csv')\n",
    "assessment_report.to_csv('./DR46/database_report/assessment_report.csv')\n",
    "subject_human_report.to_csv('./DR46/database_report/subject_human_report.csv')\n",
    "subject_organism_report.to_csv('./DR46/database_report/subject_organism_report.csv')\n",
    "experiment_sample_report.to_csv('./DR46/database_report/experiment_sample_report.csv')\n",
    "treatment_report.to_csv('./DR46/database_report/treatment_report.csv')\n",
    "STUDY_report.to_csv('./DR46/database_report/STUDY_report.csv')\n",
    "COD_report.to_csv('./DR46/database_report/COD_report.csv')\n",
    "CAT_report.to_csv('./DR46/database_report/CAT_report.csv')\n",
    "IE_report.to_csv('./DR46/database_report/IE_report.csv')\n",
    "AOC_report.to_csv('./DR46/database_report/AOC_report.csv')\n",
    "PERSONNEL_report.to_csv('./DR46/database_report/PERSONNEL_report.csv')\n",
    "PV_report.to_csv('./DR46/database_report/PV_report.csv')\n",
    "FILE_report.to_csv('./DR46/database_report/FILE_report.csv')\n",
    "LINK_report.to_csv('./DR46/database_report/LINK_report.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "ced6de84",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Counter({'Multiplex Bead Array Assay': 2,\n",
       "         'ELISA': 296,\n",
       "         'Flow Cytometry': 51,\n",
       "         'Neutralizing Antibody Titer Assay': 26,\n",
       "         'Chemiluminescent Assay': 22,\n",
       "         'Pseudovirus Neutralization Assay': 59,\n",
       "         'ELISPOT': 25,\n",
       "         'Microscopy': 18,\n",
       "         'RNA Sequencing': 4,\n",
       "         'PCR': 143,\n",
       "         'Multiplex Immunoassay': 11,\n",
       "         'T cell receptor repertoire sequencing assay': 8,\n",
       "         'Immunohistochemistry': 8,\n",
       "         'Western Blot': 2,\n",
       "         'Lateral Flow Assay': 20})"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from collections import Counter \n",
    "Counter(experiment_sample_report['Experiment Name'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "d1325bd6",
   "metadata": {},
   "outputs": [],
   "source": [
    "a = [\n",
    "    'CA260476', '21X09175N91019D00024', 'CA260517', 'CA260476', 'CA260476', 'CA260582', 'CA260476', \n",
    "    'CA260582', 'CA260591', 'CA260582', 'CA261277', 'CA260581', 'CA260563', 'CA260492', 'CA260476', \n",
    "    'CA260539', 'CA260581', 'CA260582', 'CA260582', 'CA260517', 'CA260492', 'CA260492', 'CA260517', \n",
    "    'CA260563', 'CA260541', 'CA260563', 'CA260476', 'CA260476', 'CA260560', 'CA260476', '21X092', 'CA260492', \n",
    "    'CA260492', 'CA260492', 'CA260492', 'CA260476', 'CA260508', 'CA260469', 'CA260591', 'CA260591', 'CA260476', \n",
    "    'CA260476', 'CA260476', 'CA260584', 'CA260476', 'CA260591', 'CA260476', 'CA260476', 'CA260560', 'CA260541', \n",
    "    'CA260476', 'CA260563', 'CA260582', 'CA260563', 'CA260560', 'CA260543', 'CA260543', 'CA260543', 'CA260582', \n",
    "    'CA260492', 'CA261277', 'CA261277', '75N91019D0002', 'Cordon-Cardo', '75N91019D00024', 'CA260560', 'CBC', \n",
    "    'CA260591', 'CA260476', 'CA260541', '21X091', 'CA260581', 'CA260560', 'CA260526', 'CA260582', 'CA260582', \n",
    "    'CA261277', 'CA261277', 'CA260476', 'CA260582', '', 'CA260476', 'CA260476', '\"', 'CA260563\"', 'CA260476', \n",
    "    'CA260541', 'CA260591', 'CA260539', 'CA260582', 'CA260517', 'CA260517', 'CA260476', 'CA260591', 'CA260591', \n",
    "    'CA260591', 'CA260543', 'CA260492', 'CA260517', 'CA260582', 'CA260517', 'CA260543', 'CA260581', 'CA260476', \n",
    "    'CA260476', 'CA260476', 'CA260582', 'CA260543', 'CA261277', 'CA261277', 'CA260507', 'CA260563', 'CA260563', \n",
    "    'CA260476', 'CA260582', 'CA260492', 'CA261277', 'CA260582', 'CA260543', 'CA260517', 'CA260543', 'CA260582', \n",
    "    'CA260476', 'CA260584', 'CA260563', 'CA260476', 'CA260476', 'CA260563', 'CA260476', 'CA260517', 'CA260581', \n",
    "    'CA260560', 'CA260476', 'CA260543', 'CA260543', 'CA260543', 'CA260543', 'CA260543', 'CA260543', 'CA260543', \n",
    "    'CA260543', 'CA260517', 'CA260539', 'CA260517', 'CA260476', 'CA260591', 'CA260591', 'CA260591', 'CA260582', \n",
    "    'CA260591', 'CA261277', 'CA261277', 'CA260492', 'CA260476', 'CA260476', 'CA260563', 'CA260476', 'CA261276', \n",
    "    'CA260563', 'CA260543', 'CA260560', 'CA260526', '21-1175', '', 'CA260582', 'CA260581', 'CA260581', 'CA261277', \n",
    "    'CA260582', 'CA260563', 'CA260476', 'CA260476', 'CA260476', 'CA260543', 'CA260508', 'CA260476', 'CA260508', 'CA260492', 'CA260476', 'CA260476', 'CA261276', 'CA260476', 'CA261277', 'CA260541', 'CA260541', 'CA260543', 'CA261277', 'CA261277', 'CA261277', 'CA261277', 'CA260588', 'CA260476', 'CA261277', 'CA260476', 'CA260476', 'CA260476', 'CA260541', 'CA260462', 'CA260476', 'CA260476', 'CA260476', 'CA260476', 'CA260476', 'CA260476', 'CA260476', 'CA260476', 'CA260560', 'CA260476', 'CA260543', 'CA260517', 'CA260492', 'CA260492', 'CA260541', 'CA260560', 'CA260591', 'CA260582', 'CA260582', 'CA260582', 'CA260492', 'CA260517', 'CA260517', 'CA260517', 'CA260539', 'CA260539', 'CA260591', 'CA260507', 'CA260541', 'CA260560', 'CA260560', 'CA260560', 'CA260560', 'CA260563', 'CA260581', 'CA260591', 'CA260591', 'CA260591', 'CA260591', 'CA260517', 'CA260541', 'CA260517', 'CA260517', 'CA260582', 'CA260476', 'CA260563', 'CA260541', 'CA261277', 'CA260469', 'CA260508', 'CA260476', 'CA260591', 'CA260526', 'CA260507', 'CA260581', 'CA260476', 'CA260476', '', 'CA260476', 'CA260560', 'CA260560', \n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "9bcbde94",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'',\n",
       " '\"',\n",
       " '21-1175',\n",
       " '21X091',\n",
       " '21X09175N91019D00024',\n",
       " '21X092',\n",
       " '75N91019D0002',\n",
       " '75N91019D00024',\n",
       " 'CA260462',\n",
       " 'CA260469',\n",
       " 'CA260476',\n",
       " 'CA260492',\n",
       " 'CA260507',\n",
       " 'CA260508',\n",
       " 'CA260517',\n",
       " 'CA260526',\n",
       " 'CA260539',\n",
       " 'CA260541',\n",
       " 'CA260543',\n",
       " 'CA260560',\n",
       " 'CA260563',\n",
       " 'CA260563\"',\n",
       " 'CA260581',\n",
       " 'CA260582',\n",
       " 'CA260584',\n",
       " 'CA260588',\n",
       " 'CA260591',\n",
       " 'CA261276',\n",
       " 'CA261277',\n",
       " 'CBC',\n",
       " 'Cordon-Cardo'}"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(a)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "258a1ebe",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
